web/tests/typeahead_helper.test.js

```diff
@@ -442,11 +442,11 @@ test("sort_recipients pm counts", () => {
     people.set_recipient_count_for_testing(b_bot.user_id, 1);
 
     assert.deepEqual(get_typeahead_result("b"), [
+        "a_bot@zulip.com",
```

Comment:
Hmm, why is `a_bot` appearing before the `b_` matches? I would have thought the higher priority for the start of string matching operations would have resolved in this being after those.

Reply:
In the set of matches returned by the altered `triage` function, start of string matches _are_ placed first, however these matches are finally sorted by pms (the function responsible is `compare_by_pms`) where the tie breaker is an alpha sort, which then puts `a_bot` first

Reply:
I think it would be good to first prioritize beginning-of-result matches before word-boundary matches. That feels more intuitive to me (and also wouldn't change this test).

Reply:
Agree. Related discussion [here](https://chat.zulip.org/#narrow/stream/9-issues/topic/stream.20typeahead.20exact.20matches/near/1520348). The implementation needs to be carefully thought out to enable retaining precedence of exact and beginning-of-result matches over word-boundary matches even after further sorting though 🤔  I'll get around to this soon, hopefully!

Reply:
That sounds good to me from a suer perspective.

Reply:
Just realised that I'd made a mistake in interpreting this new sorting order here.... `a_bot` is at the top since it has a higher pm_count (not coz of alpha sort). However, if it had the same pm_count as, say `b_bot`, it'd still be put before it since in this case, alpha sort is used as a tie breaker. FYI @timabbott

Reply:
Finally got around to fixing the issue of recipient sorting. Since in this case the sorting post triage is incredibly complex - multistep and dependent on several factors - there was no way to reduce it to a simple sorting comparator that could be passed to `triage`. 

Instead I changed the `triage` function to return the word boundary matches separately from the regular matches (exact or beginning matches) for the case of recipients, and then in the caller, the word boundary matches are sorted separately from the regular matches and prioritised right after them. Though this does make the triage function a little clunky, the sorting algorithms are not tampered with, and the final ordering is as should be.

Reply:
I think the logic is good, posted a suggestion for how to make the API cleaner with the same effect.